misk/src/main/kotlin/misk/digester/SlidingWindowDigest.kt

```diff
@@ -0,0 +1,182 @@
+) {
+
+  internal val windows: MutableList<WindowDigest> = mutableListOf()
+  /**
+   * Adds the given value to all currently open t-openDigests.
+   * It is important to note that an observed value is not immediately
+   * reflected in calls to Quantile.
+   */
+  @Synchronized fun observe(value: Double) {
+    for (digest in openDigests(true)) {
+      digest.Digest.add(value)
+    }
+  }
+
+  /**
+   * Returns estimated value for a quantile. The returned value
+   * may not include recently observed values due to how sliding windows are approximated.
+   * If no data has been observed then NaN is returned.
+   */
+  @Synchronized fun quantile(quantile: Double): Double {
+    val now = ZonedDateTime.now(utcNowClock)
+    for (window in windows.reversed()) {
+      if (!now.isBefore(window.window.end)) {
+        return window.Digest.quantile(quantile)
+      }
+    }
+
+    return Double.NaN
+  }
+
+  /**
+   * Returns a snapshot of estimated values for quantiles, along with the count of observations and their sum.
+   * The returned values may not include recent observations due to how sliding windows are approximated.
+   * If no data has been observed then a slice of NaNs of having len(quantiles) is returned and NaN is returned for the sum.
+   */
+  @Synchronized fun snapshot(quantiles: List<Double>): Snapshot {
+    val now = ZonedDateTime.now(utcNowClock)
+    for (i in windows.count() - 1 downTo 0) {
+      if (!now.isBefore(windows[i].window.end)) {
+        val quantileVals = DoubleArray(quantiles.count()) { Double.NaN }
+        val digest = windows[i].Digest
+        quantiles.forEachIndexed { ii, quantile ->
+          quantileVals[ii] = digest.quantile(quantile)
+        }
+        return Snapshot(
+            quantileVals,
+            digest.count(),
+            digest.sum()
+        )
+      }
+    }
+    return Snapshot(
+        DoubleArray(quantiles.count()) { Double.NaN },
+        0,
+        Double.NaN
+    )
+  }
+
+  /**
+   * Returns all WindowDigests that ended starting from the given time (inclusive).
+   *The returned WindowDigest are ordered by their start time.
+   */
+  @Synchronized fun closedDigests(from: ZonedDateTime): List<WindowDigest> {
+    deleteOlderDigests()
+    val wds: MutableList<WindowDigest> = mutableListOf()
+    for (wd in windows) {
+      if (!from.isAfter(wd.window.end)) {
+        wds.add(wd)
+      }
+    }
+    return wds.toList()
+  }
+
+  /**
+   * Merges in the data from the given WindowDigests.
+   * The given windowDigests should use the same windowing boundaries
+   * as this; if they do not then quantiles reported by this sliding
+   * window digest may be incorrect.
+   */
+  @Synchronized fun mergeIn(windowDigests: List<WindowDigest>) {
+    deleteOlderDigests()
+    for (wd in windowDigests) {
+      var found = false
+      for (existing in windows) {
+        if (existing.window.equals(wd.window)) {
+          wd.Digest.mergeInto(existing.Digest)
```

Comment:
@alecholmes The changes you are proposing would break this call as the Digest here has an abstract type of `TDigest<*>` so the mergeInto does not know which type it can allow to merge. I think we might need the way we abstract in this case?